File: ui-ngx/src/app/core/http/role.service.ts
///
/// Copyright © 2016-2025 The Thingsboard Authors
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.
///

import { Injectable } from '@angular/core';
import { defaultHttpOptionsFromConfig, RequestConfig } from './http-utils';
import { Role, Permission, UserRole, RoleTemplate } from '@shared/models/role.models';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { PageLink } from '@shared/models/page/page-link';
import { PageData } from '@shared/models/page/page-data';

@Injectable({
  providedIn: 'root'
})
export class RoleService {

  constructor(
    private http: HttpClient
  ) { }

  public getRoles(pageLink: PageLink, config?: RequestConfig): Observable<PageData<Role>> {
    return this.http.get<PageData<Role>>(`/api/roles${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getRole(roleId: string, config?: RequestConfig): Observable<Role> {
    return this.http.get<Role>(`/api/role/${roleId}`, defaultHttpOptionsFromConfig(config));
  }

  public getRoleByName(name: string, config?: RequestConfig): Observable<Role> {
    return this.http.get<Role>(`/api/role/name/${encodeURIComponent(name)}`,
      defaultHttpOptionsFromConfig(config));
  }

  public saveRole(role: Role, config?: RequestConfig): Observable<Role> {
    return this.http.post<Role>('/api/role', role, defaultHttpOptionsFromConfig(config));
  }

  public deleteRole(roleId: string, config?: RequestConfig): Observable<void> {
    return this.http.delete<void>(`/api/role/${roleId}`, defaultHttpOptionsFromConfig(config));
  }

  public getSystemRoles(config?: RequestConfig): Observable<Role[]> {
    return this.http.get<Role[]>('/api/roles/system', defaultHttpOptionsFromConfig(config));
  }

  // Permission endpoints
  public getRolePermissions(roleId: string, config?: RequestConfig): Observable<Permission[]> {
    return this.http.get<Permission[]>(`/api/role/${roleId}/permissions`,
      defaultHttpOptionsFromConfig(config));
  }

  public savePermission(permission: Permission, config?: RequestConfig): Observable<Permission> {
    return this.http.post<Permission>('/api/permission', permission, defaultHttpOptionsFromConfig(config));
  }

  public deletePermission(permissionId: string, config?: RequestConfig): Observable<void> {
    return this.http.delete<void>(`/api/permission/${permissionId}`, defaultHttpOptionsFromConfig(config));
  }

  public deleteRolePermissions(roleId: string, config?: RequestConfig): Observable<void> {
    return this.http.delete<void>(`/api/role/${roleId}/permissions`, defaultHttpOptionsFromConfig(config));
  }

  // User-Role endpoints
  public assignRoleToUser(userId: string, roleId: string, config?: RequestConfig): Observable<UserRole> {
    return this.http.post<UserRole>(`/api/user/${userId}/role/${roleId}`, null,
      defaultHttpOptionsFromConfig(config));
  }

  public unassignRoleFromUser(userId: string, roleId: string, config?: RequestConfig): Observable<void> {
    return this.http.delete<void>(`/api/user/${userId}/role/${roleId}`, defaultHttpOptionsFromConfig(config));
  }

  public getUserRoles(userId: string, config?: RequestConfig): Observable<Role[]> {
    return this.http.get<Role[]>(`/api/user/${userId}/roles`, defaultHttpOptionsFromConfig(config));
  }

  public getRoleUsers(roleId: string, pageLink: PageLink, config?: RequestConfig): Observable<PageData<UserRole>> {
    return this.http.get<PageData<UserRole>>(`/api/role/${roleId}/users${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config));
  }

  public getUserPermissions(userId: string, config?: RequestConfig): Observable<Permission[]> {
    return this.http.get<Permission[]>(`/api/user/${userId}/permissions`,
      defaultHttpOptionsFromConfig(config));
  }

  public checkUserPermission(userId: string, resource: string, operation: string,
                             config?: RequestConfig): Observable<boolean> {
    return this.http.get<boolean>(`/api/user/${userId}/permission/check`,
      {
        ...defaultHttpOptionsFromConfig(config),
        params: { resource, operation }
      });
  }

  // Role Template endpoints
  public getRoleTemplates(category?: string, config?: RequestConfig): Observable<RoleTemplate[]> {
    let url = '/api/role/templates';
    if (category) {
      url += `?category=${encodeURIComponent(category)}`;
    }
    return this.http.get<RoleTemplate[]>(url, defaultHttpOptionsFromConfig(config));
  }

  public createRoleFromTemplate(templateId: string, roleName: string,
                                config?: RequestConfig): Observable<Role> {
    return this.http.post<Role>(`/api/role/template/${templateId}/create`,
      { name: roleName },
      defaultHttpOptionsFromConfig(config));
  }

  public saveRoleAsTemplate(roleId: string, template: RoleTemplate,
                            config?: RequestConfig): Observable<RoleTemplate> {
    return this.http.post<RoleTemplate>(`/api/role/${roleId}/template`, template,
      defaultHttpOptionsFromConfig(config));
  }
}

File: common/dao-api/src/main/java/org/thingsboard/server/dao/role/RoleService.java
/**
 * Copyright © 2016-2025 The Thingsboard Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.thingsboard.server.dao.role;

import org.thingsboard.server.common.data.id.RoleId;
import org.thingsboard.server.common.data.id.TenantId;
import org.thingsboard.server.common.data.id.UserId;
import org.thingsboard.server.common.data.page.PageData;
import org.thingsboard.server.common.data.page.PageLink;
import org.thingsboard.server.common.data.permission.Permission;
import org.thingsboard.server.common.data.permission.Role;
import org.thingsboard.server.common.data.permission.UserRole;
import org.thingsboard.server.dao.entity.EntityDaoService;

import java.util.List;

public interface RoleService extends EntityDaoService {

    /**
     * Find role by id.
     *
     * @param tenantId the tenant id
     * @param roleId the role id
     * @return the role
     */
    Role findRoleById(TenantId tenantId, RoleId roleId);

    /**
     * Find role by tenant id and name.
     *
     * @param tenantId the tenant id
     * @param name the role name
     * @return the role
     */
    Role findRoleByTenantIdAndName(TenantId tenantId, String name);

    /**
     * Save or update role.
     *
     * @param tenantId the tenant id
     * @param role the role
     * @return saved role
     */
    Role saveRole(TenantId tenantId, Role role);

    /**
     * Delete role.
     *
     * @param tenantId the tenant id
     * @param role the role
     */
    void deleteRole(TenantId tenantId, Role role);

    /**
     * Find roles by tenant id.
     *
     * @param tenantId the tenant id
     * @param pageLink the page link
     * @return page of roles
     */
    PageData<Role> findRolesByTenantId(TenantId tenantId, PageLink pageLink);

    /**
     * Find all system roles.
     *
     * @return list of system roles
     */
    List<Role> findSystemRoles();

    /**
     * Find custom roles by tenant id.
     *
     * @param tenantId the tenant id
     * @return list of custom roles
     */
    List<Role> findCustomRolesByTenantId(TenantId tenantId);

    /**
     * Assign role to user.
     *
     * @param userId the user id
     * @param roleId the role id
     * @return user role
     */
    UserRole assignRoleToUser(UserId userId, RoleId roleId);

    /**
     * Unassign role from user.
     *
     * @param userId the user id
     * @param roleId the role id
     */
    void unassignRoleFromUser(UserId userId, RoleId roleId);

    /**
     * Find roles by user id.
     *
     * @param userId the user id
     * @return list of roles
     */
    List<Role> findRolesByUserId(UserId userId);

    /**
     * Find user roles by user id.
     *
     * @param userId the user id
     * @return list of user roles
     */
    List<UserRole> findUserRolesByUserId(UserId userId);

    /**
     * Add permission to role.
     *
     * @param roleId the role id
     * @param resource the resource
     * @param operation the operation
     * @return permission
     */
    Permission addPermissionToRole(RoleId roleId, String resource, String operation);

    /**
     * Remove permission from role.
     *
     * @param roleId the role id
     * @param resource the resource
     * @param operation the operation
     */
    void removePermissionFromRole(RoleId roleId, String resource, String operation);

    /**
     * Find permissions by role id.
     *
     * @param roleId the role id
     * @return list of permissions
     */
    List<Permission> findPermissionsByRoleId(RoleId roleId);

    /**
     * Find all permissions for user (from all user's roles).
     *
     * @param userId the user id
     * @return list of permissions
     */
    List<Permission> findPermissionsByUserId(UserId userId);

    /**
     * Check if user has permission.
     *
     * @param userId the user id
     * @param resource the resource
     * @param operation the operation
     * @return true if user has permission
     */
    boolean hasPermission(UserId userId, String resource, String operation);

}
